src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -104,99 +102,104 @@ def to_json_dict(self) -> dict:
             "version": self.version,
             "packaging": self.packaging,
         }
-        if self.url:
-            ret["url"] = self.url
         return ret
 
     @classmethod
     def from_coord_str(cls, s: str) -> Coordinate:
-        # if there's a `url=` part, capture the URL after it (otherwise a no-op):
-        s, _, quoted_url = s.partition(",url=")
-        url = url_unquote(quoted_url)  # no effect if `quoted_url` is empty
-
         parts = s.split(":")
         return cls(
             group=parts[0],
             artifact=parts[1],
             version=parts[2],
             packaging=parts[3] if len(parts) == 4 else "jar",
-            url=url if url else None,
         )
 
+    def as_requirement(self) -> ArtifactRequirement:
+        """Creates a `RequirementCoordinate` from a `Coordinate`."""
+        return ArtifactRequirement(coordinate=self)
+
     def to_coord_str(self, versioned: bool = True) -> str:
         unversioned = f"{self.group}:{self.artifact}"
         version_suffix = ""
-        url_suffix = ""
         if versioned:
             version_suffix = f":{self.version}"
-        if self.url:
-            url_suffix = f",url={url_quote_plus(self.url)}"
-        return f"{unversioned}{version_suffix}{url_suffix}"
+        return f"{unversioned}{version_suffix}"
 
-    @staticmethod
-    def from_jvm_artifact_target(target: Target) -> Coordinate:
+
+class Coordinates(DeduplicatedCollection[Coordinate]):
+    """An ordered list of `Coordinate`s."""
+
+
+@dataclass(frozen=True)
+class ArtifactRequirement:
+    """A single Maven-style coordinate for a JVM dependency, along with information of how to fetch
+    the dependency if it is not to be fetched from a Maven repository."""
+
+    coordinate: Coordinate
+
+    url: str | None = None
+    jar: JvmArtifactJarSourceField | None = None
```

Comment:
Maybe a useful case for a three way union of `str | JvmArtifactJarSourceField | None` ?

Reply:
What's the intention there?

Reply:
Improved type safety, since it isn't legal to have both of them at once. On the rust side, this case would be handled via an enum to represent the three different valid states.

Reply:
Oh, you mean to merge `url` and `jar? Let me think about how to make that one work